test/distributed/org/apache/cassandra/distributed/test/ReplicaFilteringProtectionTest.java

```diff
@@ -0,0 +1,139 @@
+    {
+        String tableName = "missed_updates_cache_warn";
+        String fullTableName = KEYSPACE + '.' + tableName;
+        cluster.schemaChange(withKeyspace("CREATE TABLE %s." + tableName + " (k int PRIMARY KEY, v text)"));
+
+        // The warning threshold provided is one less than the total number of rows returned
+        // to the coordinator from all replicas and therefore should be triggered but not fail the query.
+        testMissedUpdates(fullTableName, REPLICAS * ROWS - 1, Integer.MAX_VALUE);
+    }
+
+    @Test
+    public void testMissedUpdatesAboveCachingFailThreshold()
+    {
+        String tableName = "missed_updates_cache_fail";
+        String fullTableName = KEYSPACE + '.' + tableName;
+        cluster.schemaChange(withKeyspace("CREATE TABLE %s." + tableName + " (k int PRIMARY KEY, v text)"));
+        try
+        {
+            // The failure threshold provided is one less than the total number of rows returned
+            // to the coordinator from all replicas and therefore should fail the query.
+            testMissedUpdates(fullTableName, 1, REPLICAS * ROWS - 1);
+        }
+        catch (RuntimeException e)
+        {
+            assertEquals(e.getCause().getClass().getName(), TooManyCachedRowsException.class.getName());
+        }
+    }
+
+    private void testMissedUpdates(String fullTableName, int warnThreshold, int failThreshold)
+    {
+        cluster.get(1).runOnInstance(() -> StorageService.instance.setCachedReplicaRowsWarnThreshold(warnThreshold));
+        cluster.get(1).runOnInstance(() -> StorageService.instance.setCachedReplicaRowsFailThreshold(failThreshold));
+
+        for (int i = 0; i < ROWS; i++)
+        {
+            cluster.coordinator(1).execute("INSERT INTO " + fullTableName + "(k, v) VALUES (?, 'old')", ConsistencyLevel.ALL, i);
+        }
+
+        Object[][] initialRows =
+            cluster.coordinator(1).execute("SELECT * FROM " + fullTableName + " WHERE v = 'old' LIMIT ? ALLOW FILTERING",
+                                           ConsistencyLevel.ALL, ROWS);
+        assertRows(initialRows,
+                   row(1, "old"), row(0, "old"), row(2, "old"));
+
+        // Update all rows on only one replica, leaving the entire dataset of the remaining replica out-of-date:
+        for (int i = 0; i < ROWS; i++)
+        {
+            cluster.get(1).executeInternal("UPDATE " + fullTableName + " SET v = 'new' WHERE k = ?", i);
+        }
+
+        Object[][] emptyOldMatches =
+            cluster.coordinator(1).execute("SELECT * FROM " + fullTableName + " WHERE v = 'old' LIMIT ? ALLOW FILTERING",
+                                           ConsistencyLevel.ALL, ROWS);
+        assertRows(emptyOldMatches);
+
+        Object[][] newMatches =
+            cluster.coordinator(1).execute("SELECT * FROM " + fullTableName + " WHERE v = 'new' LIMIT ? ALLOW FILTERING",
+                                           ConsistencyLevel.ALL, ROWS);
+        assertRows(newMatches,
```

Comment:
We can't know which query between the one for `old` and the one for `new` is triggering the threshold.

Reply:
They'll fail exactly the same way, so it's alway the first. I figured it would be okay to leave it this way, given we want to verify both for the non-failing cases. WDYT?